### PR TITLE
Support Pads for MaxPoolSingleOut

### PR DESCRIPTION
**Pull request by tungld**
This patch is to use declarative rewriting rules to rewrite ONNXMaxPoolSingleOutOp with pads into composition of ONNXPadConstantValuePadOp and ONNXMaxPoolSingleOut with pads = [0, 0, .., 0].

### CODE REVIEW

**Comment by tjingrant** [comment]
I'm not exactly sure why we are building padded operations with explicit padding followed by a non-padded operation. I asked about the bigger picture in this thread https://github.com/clang-ykt/ONNF/pull/82#issuecomment-586693428 but didn't get an answer. Could you explain to me what's the goal here? Reposting my question below

> @doru1004 @AlexandreEichenberger I'm a bit curious what's next for this thread of work (handling padding in convolution). Per our discussion, we are to merge padding + convolution at some point later in our pipeline so that padding becomes implicit, which is good for performance. Then what compilation process will specifically benefit from the separation of padding and convolution?
> 
> It would be nice for me to have some clues about the bigger picture.

 cc @doru1004 @AlexandreEichenberger

**Comment by tungld** [comment]
In my understanding, for now this is to simplify the handling of multiple cases of padding using krnl ops when lowering MaxPool or Conv. Of course, handling padding inside the lowering code would be better for performance. 

I’d love to hear comments from Alex and Doru on this also.

**Comment by doru1004** [comment]
Hi Tung, I don't see a need for a new op here. We already have the means to express this transformation with existing operations.

```
MaxPool with pads = [a,b,c,d]
```
=>
```
Explicit padding op with pads = [a,b,c,d]
+
MaxPool with pads = [0,0,0,0] and auto_pad set to NOTSET.
```

**Comment by tungld** [comment]
The reason is that I was not able to rewrite MaxPoolSingleOut from MaxPoolSingleOut in this pass. I have tried to change legal/illegal targets for the pass, but I haven't succeeded.

We can rewrite MaxPoolSingleOut from MaxPoolSingleOut if we move these declarative rewriting rules into canonicalization pass. However, canonicalization is done after shape inference, so we need to infer a result shape for the newly added padding op. Let me try to see if I can proceed in this way.

**Comment by tungld** [comment]
> Hi Tung, I don't see a need for a new op here. We already have the means to express this transformation with existing operations.
> 
> ```
> MaxPool with pads = [a,b,c,d]
> ```
> 
> =>
> 
> ```
> Explicit padding op with pads = [a,b,c,d]
> +
> MaxPool with pads = [0,0,0,0] and auto_pad set to NOTSET.
> ```

@doru1004, I just rewrite MaxPool as the above by using canonicalization pass, but do not infer shape for the result of the explicit padding op, just returning UnrankedTensorType. It is better to call shape-inference pass again after canonicalization, which reduces burden to infer shape in canonicalization pass and also avoids duplication code for shape inference here and there. If you agree on this, we could create a new PR to call shape-inference pass after canonicalization or we could do it in this patch.

**Comment by doru1004** [comment]
> @doru1004, I just rewrite MaxPool as the above by using canonicalization pass, but do not infer shape for the result of the explicit padding op, just returning UnrankedTensorType. It is better to call shape-inference pass again after canonicalization, which reduces burden to infer shape in canonicalization pass and also avoids duplication code for shape inference here and there. If you agree on this, we could create a new PR to call shape-inference pass after canonicalization or we could do it in this patch.

It seems to me that we would benefit from separating rank inference from actual shape inference. If we had a rank inference pass we would be able to infer the rank, perform canonicalization and then call shape inference. What do you think?

**Comment by tungld** [comment]
Though it is easy to compute the output shape of the newly added constant padding operation, the newly added operation already has its inferShape method. Hence, if we compute the output shape again without using its inferShape method will potentially causes problem, and this way does not scale well to other declarative rewriting rules we may have in the future.

What I mean is this order of passes: 
- shape inference pass
- canonicalization
- shape inference pass

> padded_shape[i] = input_shape[i] + pads[i] + pads[i + rank]

What happens if  `input_shape[i] = -1`?

**Comment by tungld** [comment]
Doru, this patch also helps replacing SplitConvOpPattern (https://github.com/onnx/onnx-mlir/blob/master/src/pass/onnx_rewrite.cpp#L36) by a declarative rewriting rule like the following one without any additional code:
```TableGen
def SplitConvOpPattern: Pat<
   (ONNXConvNoBiasOp:$res
      $x,
      $w, $auto_pad, $dilation, $group, $kernel_shape,
      $pads,
      $strides),
   (ONNXConvNoBiasOp
      (ONNXPadConstantValuePadOp $x,
         (PadArrayAttrWithZeroTrail<2> $pads),
         (FloatAttrOfValue<0> $res),
         (StringAttrOfValue<"constant">)),
      $w, $auto_pad, $dilation, $group, $kernel_shape,
      (ArrayAttrOfZeroWithTrail<2> $pads),
      $strides),
   [(HasNonZeroInArrayAttr:$pads)]
 >;
```

**Comment by doru1004** [comment]
> * shape inference pass
> * canonicalization
> * shape inference pass

I agree with this sequence but I can't see it in the diff. Is it not needed in the latest version?

File: src/pass/onnx_rewrite.td [hunk 1]
```diff
@@ -24,4 +24,54 @@ include "dialect/onnx/onnx.td"
...
+// This function is used for padding attribute in MaxPoolSingleOut.
+class ArrayAttrOfZeroWithTrail<int trailCount> : NativeCodeCall<"createArrayAttrOfZeroWithTrail($_builder, $0, " # trailCount # ")">;
+
+// Pad a ArrayAttr with trails of zeros.
+// This function is used for padding attribute in MaxPoolSingleOut.
+class PadArrayAttrWithZeroTrail<int trailCount> : NativeCodeCall<"padArrayAttrWithZeroTrail($_builder, $0," # trailCount # ")">;
+
+// Check whether an ArrayAttr contains non-zero values or not.
+def HasNonZeroInArrayAttr : Constraint<CPred<"hasNonZeroInArrayAttr($_self)">, "has non-zero elements">;
+def HasAllZeroInArrayAttr : Constraint<CPred<"!hasNonZeroInArrayAttr($_self)">, "has non-zero elements">;
```
Comment (doru1004) [inline src/pass/onnx_rewrite.td]: This doesn't seem to be used anywhere.

File: src/pass/onnx_rewrite.cpp [hunk 2]
```diff
@@ -17,6 +17,46 @@
...
+ArrayAttr createArrayAttrOfZeroWithTrail(
+    PatternRewriter &rewriter, ArrayAttr origAttrs, int trailCount) {
+  int nElements = origAttrs.getValue().size() + trailCount * 2;
+  SmallVector<int64_t, 4> vals(nElements, 0);
+  return rewriter.getI64ArrayAttr(vals);
+}
+
+// Pad a ArrayAttr with trails of zeros.
+// This function is used for padding attribute in MaxPoolSingleOut.
+ArrayAttr padArrayAttrWithZeroTrail(
```
Comment (doru1004) [inline src/pass/onnx_rewrite.cpp]: I would rename this to:

insertZerosForNonPaddedDims

And then I would insert a more extensive comment explaining how the zeros are inserted:

```
pads = [B1, B2, ... Bk, E1, E2, ..., Ek]

becomes:

pads = [0,... 0, B1, B2, ... Bk, 0,... 0, E1, E2, ..., Ek]
        |_____|                  |_____|
        nZeros                    nZeros
```

File: src/pass/onnx_rewrite.cpp [hunk 3]
```diff
@@ -17,6 +17,46 @@
...
+        break;
+      }
+    }
+  }
+  return !allZeros;
+}
+
+// Create an ArrayAttr of IntergerAttr(s) of zero values.
+// This function is used for padding attribute in MaxPoolSingleOut.
+ArrayAttr createArrayAttrOfZeroWithTrail(
```
Comment (doru1004) [inline src/pass/onnx_rewrite.cpp]: "Trail" suggests to me that the zeros are added at the end. I would just call this method createExtendedArrayAttrOfZeros

File: src/pass/onnx_rewrite.cpp [hunk 4]
```diff
@@ -17,6 +17,46 @@
...
+    PatternRewriter &rewriter, ArrayAttr origAttrs, int trailCount) {
```
Comment (doru1004) [inline src/pass/onnx_rewrite.cpp]: Instead of trailCount I would use extensionLength

File: test/mlir/onnx/onnx_canonicalization.mlir [hunk 5]
```diff
@@ -77,3 +77,23 @@ func @test_gemm_add_fusion_rank3(%arg0: tensor<128x128x256xf32>, %arg1: tensor<1
   // return [[GEMM]] : tensor<*xf32>
 }
 
+//CHECK-LABEL: @test_maxpoolsingleout_split(%{{.*}}: tensor<5x5x32x32xf32>) -> tensor<5x8x32x39xf32> {
+func @test_maxpoolsingleout_split(%arg0: tensor<5x5x32x32xf32>) -> tensor<5x8x32x39xf32> {
+  %0 = "onnx.MaxPoolSingleOut"(%arg0) {auto_pad = "NOTSET", ceil_mode = 0, kernel_shape = [5,3], pads = [1, 2, 3, 4] } : (tensor<5x5x32x32xf32>) -> tensor<5x8x32x39xf32>
+  "std.return"(%0) : (tensor<5x8x32x39xf32>) -> ()
+
+  // CHECK-NEXT: %0 = "onnx.PadConstantValuePad"(%arg0) {constant_value = 0.000000e+00 : f32, mode = "constant", pads = [0, 0, 1, 2, 0, 0, 3, 4]} : (tensor<5x5x32x32xf32>) -> tensor<5x8x32x39xf32>
+  // CHECK-NEXT: %1 = "onnx.MaxPoolSingleOut"(%0) {auto_pad = "NOTSET", ceil_mode = 0 : i64, kernel_shape = [5, 3], pads = [0, 0, 0, 0, 0, 0, 0, 0], storage_order = 0 : i64} : (tensor<5x8x32x39xf32>) -> tensor<5x8x32x39xf32>
```
Comment (doru1004) [inline test/mlir/onnx/onnx_canonicalization.mlir]: So it looks like the length of the pads in the case of MaxPool has entries for all input dimensions i.e. 4 dimensions. In the spec the list of pads only handles the spatial dimensions. We may run into trouble when ingesting models which only pad the spatial dimensions.

File: src/pass/onnx_rewrite.td [hunk 6]
```diff
@@ -24,4 +24,54 @@ include "dialect/onnx/onnx.td"
...
+def MaxPoolSingleOutOpPaddingPattern: Pat<
+  (ONNXMaxPoolSingleOutOp:$res
+     $x,
+     $auto_pad, $ceil_mode, $dilation, $kernel_shape,
+     $pads,
+     $storage_order, $strides),
+  (ONNXMaxPoolSingleOutOp
+     (ONNXPadConstantValuePadOp $x,
+        (PadArrayAttrWithZeroTrail<2> $pads),
+        (FloatAttrOfValue<0> $res),
```
Comment (doru1004) [inline src/pass/onnx_rewrite.td]: Does this $res have anything to do with the $res at line 61? If not can we name it differently? The name of this attribute in the definition is constant_value so maybe we can give it the same name.

File: src/pass/onnx_rewrite.td [hunk 7]
```diff
@@ -24,4 +24,54 @@ include "dialect/onnx/onnx.td"
...
+        (StringAttrOfValue<"constant">)),
+     $auto_pad, $ceil_mode, $dilation, $kernel_shape,
+     (ArrayAttrOfZeroWithTrail<2> $pads),
```
Comment (doru1004) [inline src/pass/onnx_rewrite.td]: There shouldn't be a 2 here.

**Comment by tungld** [comment]
> > * shape inference pass
> > * canonicalization
> > * shape inference pass
> 
> I agree with this sequence but I can't see it in the diff. Is it not needed in the latest version?

Yes, it is still needed because the build method of ONNXPadConstantValuePadOp may return an UnrankedTensorType. I just added a shape inference after canonicalization. This makes us easily write declarative rewriting rules in the future where it is enough for newly added ops (such as ONNXPadConstantValuePadOp) to return a UnrankedTensorType instead of infer an exact shape in the rewriting rule.

File: test/mlir/onnx/onnx_canonicalization.mlir (hunk 5 above)
Comment (tungld) [reply test/mlir/onnx/onnx_canonicalization.mlir]: Fixed this. Pads are for the spatial dimensions only. Thanks for pointing out this.

**Comment by tungld** [comment]
Doru, I updated this patch according to your comments! Thank you!

**Pull request merged**